firedrake/adjoint_utils/function.py

```diff
@@ -315,25 +308,23 @@ def _ad_function_space(self, mesh):
         return self.ufl_function_space()
 
     def _reduce(self, r, r0):
-        vec = self.vector().get_local()
+        vec = self.dat.data_ro
         for i in range(len(vec)):
             r0 = r(vec[i], r0)
         return r0
 
     def _applyUnary(self, f):
-        vec = self.vector()
-        npdata = vec.get_local()
+        npdata = self.dat.data_ro
         for i in range(len(npdata)):
             npdata[i] = f(npdata[i])
-        vec.set_local(npdata)
+        self.dat.data_wo[...] = npdata
 
     def _applyBinary(self, f, y):
-        vec = self.vector()
-        npdata = vec.get_local()
-        npdatay = y.vector().get_local()
+        npdata = self.dat.data_ro
+        npdatay = y.dat.data_ro
         for i in range(len(npdata)):
             npdata[i] = f(npdata[i], npdatay[i])
-        vec.set_local(npdata)
+        self.dat.data_wo[...] = npdata
```

Comment:
`npdata` points to the read-only array on line 323, but is written into on line 326. Is doing the right thing?
If you just write into `self.dat.data_wo` directly then you can also remove the additional copy on line 327.

Reply:
done